compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,284 @@
+func (c *CachingScanner) validateBackoff(backoff time.Duration) time.Duration {
+	if backoff > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return backoff
+}
+
+func readInventoryWrap(path string) *inventoryWrap {
+	cacheContents, err := os.ReadFile(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No node scan cache file found, will run a new scan")
+			return nil
+		}
+		log.Warnf("Unable to read node scan cache, will run a new scan. Error: %v", err)
+		return nil
+	}
+
+	// deserialize stored inventory
+	var wrap inventoryWrap
+	if err := json.Unmarshal(cacheContents, &wrap); err != nil {
+		log.Warnf("Unable to load node scan cache contents, will run a new scan. Error: %v", err)
+		return nil
+	}
+	return &wrap
+}
+
+func writeInventoryWrap(w inventoryWrap, path string) error {
+	jsonWrap, err := json.Marshal(&w)
+	if err != nil {
+		return err
+	}
+
+	if err := os.WriteFile(path, jsonWrap, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// collectInventory scans the current node and returns the results as storage.NodeInventory object
+func collectInventory(nodeName string) (*storage.NodeInventory, error) {
+	metrics.ObserveScansTotal(nodeName)
+	startTime := time.Now()
+
+	// uncertifiedRHEL is set to false, as scans are only supported on RHCOS for now,
+	// which only exists in certified versions
+	componentsHost, err := nodes.Analyze(nodeName, "/host/", nodes.AnalyzeOpts{UncertifiedRHEL: false, IsRHCOSRequired: true})
+
+	scanDuration := time.Since(startTime)
+	metrics.ObserveScanDuration(scanDuration, nodeName, err)
+	log.Debugf("Collecting Node Inventory took %f seconds", scanDuration.Seconds())
+
+	if err != nil {
+		log.Errorf("Error scanning node /host inventory: %v", err)
+		return nil, err
+	}
+	log.Debugf("Components found under /host: %v", componentsHost)
+
+	protoComponents := protoComponentsFromScanComponents(componentsHost)
+
+	if protoComponents == nil {
+		log.Warn("Empty components returned from NodeInventory")
+	} else {
+		log.Infof("Node inventory has been built with %d packages and %d content sets",
+			len(protoComponents.GetRhelComponents()), len(protoComponents.GetRhelContentSets()))
+	}
+
+	// uncertifiedRHEL is false since scanning is only supported on RHCOS for now,
+	// which only exists in certified versions. Therefore, no specific notes needed
+	// if uncertifiedRHEL can be true in the future, we can add Note_CERTIFIED_RHEL_SCAN_UNAVAILABLE
+	m := &storage.NodeInventory{
+		NodeId:     uuid.Nil.String(), // The NodeID is not available in compliance, but only on Sensor and later on
+		NodeName:   nodeName,
+		ScanTime:   timestamp.TimestampNow(),
+		Components: protoComponents,
+		Notes:      []storage.NodeInventory_Note{storage.NodeInventory_LANGUAGE_CVES_UNAVAILABLE},
+	}
+
+	metrics.ObserveNodeInventoryScan(m)
+	return m, nil
+}
+
+// TODO(ROX-14029): Move conversion function into Sensor
```

Comment:
Sensor or Scanner?

```suggestion
// TODO(ROX-14029): Move conversion function into Scanner
```

Note that ROX-14029 also says "Sensor".

Reply:
Why move to Scanner: for container separation :-)